src/grid.rs

```diff
@@ -208,98 +432,207 @@ impl<T> Grid<T> {
+                self.cur.line -= AbsoluteLine(1);
                 self.cur.col = num_cols - Column(1);
-                Some(&self.grid[self.cur.line][self.cur.col])
+                Some(&self.grid.get_absolute_line_unchecked(self.cur.line)[self.cur.col])
             },
             _ => {
                 self.cur.col -= Column(1);
-                Some(&self.grid[self.cur.line][self.cur.col])
+                Some(&self.grid.get_absolute_line_unchecked(self.cur.line)[self.cur.col])
             }
         }
     }
 }
 
+/// Row indexing for the Grid
 impl<T> Index<index::Line> for Grid<T> {
     type Output = Row<T>;
 
     #[inline]
     fn index(&self, index: index::Line) -> &Row<T> {
-        &self.raw[index.0]
+        let index = self.active_to_raw_index(index);
+        &self.raw[index]
     }
 }
 
 impl<T> IndexMut<index::Line> for Grid<T> {
     #[inline]
     fn index_mut(&mut self, index: index::Line) -> &mut Row<T> {
-        &mut self.raw[index.0]
+        let index = self.active_to_raw_index(index);
+        &mut self.raw[index]
     }
 }
 
+/// Row slicing for the Grid.
+impl<T> TakeSlice<Row<T>, Line> for Grid<T> {
+    fn len(&self) -> Line {
+        self.num_lines()
+    }
+}
+
+/// Absolute slice for Grid<T>
+impl<T> Grid<T> {
+    /// Most the other functions of the grid deal only with the 'active region' of the buffer,
+    /// which basically is the area of the buffer that is still being manipulated by apps,
+    /// and isn't stored for the sole purpose of scrollback.
+    ///
+    /// This function allows access to a region in the buffer, indexed by an `AbsoluteLine`
+    /// (ie: 0 means the very oldest line of scrollback). Note that this is read-only access,
+    /// presumably for rendering purposes - the scrollback-area of the buffer
+    /// should never be modified.
+    pub fn get_absolute_region(&self, region: Range<AbsoluteLine>) -> owned_slice::Slice<VecDeque<Row<T>>, usize, Row<T>> {
+        let region = self.absolute_to_raw_index(region.start)..self.absolute_to_raw_index(region.end);
+        owned_slice::Slice::new(&self.raw, region)
+    }
+
+    /// This function allows access to a specific line in the buffer, indexed by an `AbsoluteLine`
+    /// (ie: 0 means the very oldest line of scrollback). Note that this is read-only access,
+    /// presumably for rendering purposes - the scrollback-area of the buffer
+    /// should never be modified.
+    pub fn get_absolute_line(&self, line: AbsoluteLine) -> Option<&Row<T>> {
+        if line < self.absolute_line_offset {
+            None
+        } else {
+            Some(&self.raw[self.absolute_to_raw_index(line)])
+        }
+    }
+
+    fn get_absolute_line_unchecked(&self, line: AbsoluteLine) -> &Row<T> {
+        &self.raw[self.absolute_to_raw_index(line)]
+    }
+
+    pub fn get_visible_line(&self, line: Line) -> &Row<T> {
+        let idx = self.absolute_to_raw_index(self.visible_region_start + line.to_absolute());
+        &self.raw[idx]
```

Comment:
It crashes for me right here:

```
alacritty git:(master) ✗ RUST_BACKTRACE=1 ./target/debug/alacritty
[unhandled osc_dispatch]: [['1','0',],['?',],] at line 784
[unhandled osc_dispatch]: [['1','1',],['?',],] at line 797
thread 'main' panicked at 'Out of bounds access', src/libcore/option.rs:839:4
stack backtrace:
   0: std::sys::imp::backtrace::tracing::imp::unwind_backtrace
   1: std::sys_common::backtrace::print
   2: std::panicking::default_hook::{{closure}}
   3: std::panicking::default_hook
   4: std::panicking::rust_panic_with_hook
   5: std::panicking::begin_panic
   6: std::panicking::begin_panic_fmt
   7: rust_begin_unwind
   8: core::panicking::panic_fmt
   9: core::option::expect_failed
  10: <core::option::Option<T>>::expect
             at /build/rust/src/rustc-1.22.1-src/src/libcore/option.rs:302
  11: <alloc::vec_deque::VecDeque<A> as core::ops::index::Index<usize>>::index
             at /build/rust/src/rustc-1.22.1-src/src/liballoc/vec_deque.rs:2365
  12: <alacritty::grid::Grid<T>>::get_visible_line
             at src/grid.rs:601
  13: <alacritty::term::RenderableCellsIter<'a> as core::iter::iterator::Iterator>::next
             at src/term/mod.rs:336
  14: alacritty::renderer::RenderApi::render_cells
             at src/renderer/mod.rs:827
  15: alacritty::display::Display::draw::{{closure}}
             at src/display.rs:364
  16: alacritty::renderer::QuadRenderer::with_api
             at src/renderer/mod.rs:658
  17: alacritty::display::Display::draw
             at src/display.rs:357
  18: alacritty::run
             at src/main.rs:202
  19: alacritty::main
             at src/main.rs:53
  20: __rust_maybe_catch_panic
  21: std::panicking::try
  22: std::rt::lang_start
  23: main
  24: __libc_start_main
  25: _start
thread 'pty reader' panicked at 'called `Result::unwrap()` on an `Err` value: EventsLoopClosed', src/libcore/result.rs:906:4
note: Some details are omitted, run with `RUST_BACKTRACE=full` for a verbose backtrace.
stack backtrace:
   0: std::sys::imp::backtrace::tracing::imp::unwind_backtrace
   1: std::sys_common::backtrace::print
   2: std::panicking::default_hook::{{closure}}
   3: std::panicking::default_hook
   4: std::panicking::rust_panic_with_hook
   5: std::panicking::begin_panic
   6: std::panicking::begin_panic_fmt
   7: rust_begin_unwind
   8: core::panicking::panic_fmt
   9: core::result::unwrap_failed
➜  alacritty git:(master) ✗
```

It seems that `idx` is out of bounds and it crashes Alacritty. I'm using i3-gaps and polybar and this happens when the terminal is focused and I reload polybar (among other cases where the terminals appear to crash at random when resizing, etc).

Reply:
There's a chance of it happening whenever the terminal resizes while scrolling is disabled.
As mentioned in my earlier comment 'the get_visible_lines' method is called here:
<https://github.com/neon64/alacritty/blob/master/src/term/mod.rs#L336>

I've successfully gotten it to not crash by skipping the line whenever the index is out of bounds, but there are more issues to this than just crashing.

Whenever making the terminal window smaller, there's also a chance of the prompt appearing again (or whatever the current bottom line of the terminal is). So there seems to be something fundamentally wrong with how this stuff is handled without scrollback.

Reply:
Yes, this bug has been in master for a long time. I experience it nearly everyday, but had never quite figured out exactly what the steps to repro were, and the workaround was just resize the window a little bigger.

Reply:
You're not talking about jwilm master, right? If you're talking about neon64's master, then it has probably always been an issue. I'm not sure what the best approach for solving it are though. Do you know anything about this @neon64?

Reply:
Okay so somehow, inside `Term::renderable_cells()` after a resize (smaller->bigger), `self.grid.raw.len() < self.grid.num_lines()`. This is a clear violation of one of the invariants of the `Grid` type, thus it's no surprise that it crashes soon after. So I'm trying to find how it got to that state now.